### ISSUE TITLE
"ASP.NET Core Blazor Server additional security scenarios" is incorrect in token-based authentication example

### ISSUE THREAD

**Issue by Rafnel** [dotnet/AspNetCore.Docs#29899]
This section of the MS docs: https://learn.microsoft.com/en-us/aspnet/core/blazor/security/server/additional-scenarios?view=aspnetcore-6.0#pass-tokens-to-a-blazor-server-app outlines an example for retrieving a user's access_token and refresh_token using the HttpContext at the top-level of the application (_Host.cshtml). After following this guide, I have found it to be incorrect. (But please let me know if I'm doing something incorrectly :D)

## Suggestion
Either remove this section or add a correction, although I'm not sure what the correction should be - I'm still not totally sure how this can be done in Blazor Server.

## Steps to Reproduce
- Set up a Blazor Server app with OpenIdConnect auth. I'm using Keycloak for authentication. Relevant portion of my Program.cs:
```
builder.Services.AddScoped<TokenProvider>();


builder.Services.AddAuthentication(auth =>
	{
		auth.DefaultAuthenticateScheme = CookieAuthenticationDefaults.AuthenticationScheme;
		auth.DefaultSignInScheme = CookieAuthenticationDefaults.AuthenticationScheme;
		auth.DefaultChallengeScheme = OpenIdConnectDefaults.AuthenticationScheme;
	})
	.AddCookie()
	.AddOpenIdConnect(OpenIdConnectDefaults.AuthenticationScheme, options =>
	{
		options.MetadataAddress = builder.Configuration["Keycloak:Metadata-Address"];
		options.Authority = builder.Configuration["Keycloak:Authority"];
		options.ClientId = builder.Configuration["Keycloak:Client-Id"];
		options.ClientSecret = builder.Configuration["Keycloak:Client-Secret"];
		options.ResponseType = OpenIdConnectResponseType.Code;

		options.SaveTokens = true;
		options.Scope.Add("openid");
		options.Scope.Add("profile");
		options.Scope.Add("offline_access");
	});
```
- Create a TokenProvider class as in the example. I also put a GUID on the TokenProvider class to show how this example doesn't work:
```
public class TokenProvider
	{
		private string? _token;
		private string _guid;


		public TokenProvider()
		{
			_guid = Guid.NewGuid().ToString();
		}

		public string? AccessToken
		{
			get
			{
				Trace.WriteLine("GUID: " + _guid);
				Trace.WriteLine("TOKEN: " + _token);
				return _token;
			}
			set => _token = value;
		}
		public string? RefreshToken;
	}
```
- Pull the user's access_token and refresh_token from their HttpContext as in the linked page in _Host.cshtml:
```
TokenProvider.AccessToken = await HttpContext.GetTokenAsync("access_token");
TokenProvider.RefreshToken = await HttpContext.GetTokenAsync("refresh_token");
```
- Put some dummy code in the App.razor's initializor ( or some other top-level component, doesn't really matter)
```
protected override void OnInitialized()
	{
		base.OnInitialized();
		var howdy = TokenProvider.AccessToken;
	}
```
Now run the application and access the site. You should see two back-to-back log statements when you load the site. One will include your access token, and one won't.
```
GUID: 8e7cc8a1-6dde-41af-b07d-1675fac1e2f7
TOKEN: eyJhb.....<abbreviated>
GUID: cf379e66-ac85-4264-8f3a-6c984bbe9a67
TOKEN: 
```
This happens because we are using the "ServerPrerendered" render-mode in our _Host.cshtml(This is the default render mode):
`<component type="typeof(App)" render-mode="ServerPrerendered"/>`
in ServerPrerendered the app will load once with just static assets and then load again once the server has responded (2 loads total). On the first load, it looks like we get our token as expected. Second load never gives you your token.

If you change the render-mode to "Server" You lose the first render and just get the post-server render, with no token returned by the HttpContext. Since TokenProvider is scoped to a single session, it also changes on load. We could hold onto the token if we change

`builder.Services.AddScoped<TokenProvider>();`
to
`builder.Services.AddSingleton<TokenProvider>();`
in Program.cs, but then all clients to the application will share a single TokenProvider and will be overwriting each other's tokens. We want a TokenProvider scoped to each user's session (circuit). When following the example, in the docs, it appears that the application will always get an empty access token.

Some references:
- https://stackoverflow.com/questions/60996170/blazor-scoped-service-initializing-twice
- https://stackoverflow.com/questions/72319872/token-is-coming-as-null-from-gettokenasync-in-blazor-server

The doc linked in this issue does not recognize this critical issue. The code provided in the doc does not work due to this.
---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: 6f8acbbf-c9d4-7041-c1a7-ec9ce8f0981b
* Version Independent ID: c98be365-408d-7ee6-cb74-14c44d01b0b8
* Content: [ASP.NET Core Blazor Server additional security scenarios](https://learn.microsoft.com/en-us/aspnet/core/blazor/security/server/additional-scenarios?view=aspnetcore-7.0)
* Content Source: [aspnetcore/blazor/security/server/additional-scenarios.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/blazor/security/server/additional-scenarios.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-security**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

**Comment by Rafnel** [comment]
Alright, after 2 days of this problem I found out the docs are actually correct. You **have to** make an InitialState (or whatever name) custom class to hold the token from _Host.cshtml, and pass it down to App.razor to set in the scoped TokenProvider. I foolishly thought this intermediary step was unnecessary and injected + assigned the tokens to the TokenProvider in _Host.cshtml. For some reason, you have to pass the tokens into App.razor and set them on the TokenProvider in App.razor. 

tl;dr: Don't skip these two sections or you'll be a sucker like me:

_Host.cshtml:
```
@{
    var tokens = new InitialApplicationState
    {
        AccessToken = await HttpContext.GetTokenAsync("access_token"),
        RefreshToken = await HttpContext.GetTokenAsync("refresh_token"),
        XsrfToken = Xsrf.GetAndStoreTokens(HttpContext).RequestToken
    };
}

<component ... param-InitialState="tokens" ... />
```
App.razor:
```
@inject TokenProvider TokenProvider

...

@code {
    [Parameter]
    public InitialApplicationState? InitialState { get; set; }

    protected override Task OnInitializedAsync()
    {
        TokenProvider.AccessToken = InitialState?.AccessToken;
        TokenProvider.RefreshToken = InitialState?.RefreshToken;
        TokenProvider.XsrfToken = InitialState?.XsrfToken;

        return base.OnInitializedAsync();
    }
}
```
